cpp/src/arrow/io/readahead.cc

```diff
@@ -0,0 +1,222 @@
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/io/readahead.h"
+#include "arrow/buffer.h"
+#include "arrow/status.h"
+#include "arrow/util/logging.h"
+
+#include <condition_variable>
+#include <cstring>
+#include <deque>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <thread>
+#include <utility>
+
+namespace arrow {
+namespace io {
+namespace internal {
+
+// ----------------------------------------------------------------------
+// ReadaheadSpooler implementation
+
+class ReadaheadSpooler::Impl {
+ public:
+  Impl(MemoryPool* pool, std::shared_ptr<InputStream> raw, int64_t read_size,
+       int32_t readahead_queue_size, int64_t left_padding, int64_t right_padding)
+      : pool_(pool),
+        raw_(raw),
+        read_size_(read_size),
+        readahead_queue_size_(readahead_queue_size),
+        left_padding_(left_padding),
+        right_padding_(right_padding) {
+    DCHECK_NE(raw, nullptr);
+    DCHECK_GT(read_size, 0);
+    DCHECK_GT(readahead_queue_size, 0);
+    io_worker_ = std::thread([&]() { WorkerLoop(); });
+  }
+
+  ~Impl() { ARROW_UNUSED(Close()); }
+
+  Status Close() {
+    std::unique_lock<std::mutex> lock(mutex_);
+    please_close_ = true;
+    io_wakeup_.notify_one();
+    // Wait for IO thread to finish
+    if (io_worker_.joinable()) {
+      lock.unlock();
+      io_worker_.join();
+      lock.lock();
+    }
+    return raw_->Close();
+  }
+
+  Status Read(ReadaheadBuffer* out) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    while (true) {
+      // Drain queue before querying other flags
+      if (buffer_queue_.size() > 0) {
+        *out = std::move(buffer_queue_.front());
+        DCHECK_NE(out->buffer, nullptr);
+        buffer_queue_.pop_front();
+        // Need to fill up queue again
+        io_wakeup_.notify_one();
+        return Status::OK();
+      }
+      if (!read_status_.ok()) {
+        // Got a read error, bail out
+        return read_status_;
+      }
+      if (eof_) {
+        // XXX maybe we need a EOF status code
+        out->buffer.reset();
+        return Status::OK();
+      }
+      // Readahead queue is empty and we're not closed yet, wait for more I/O
+      io_progress_.wait(lock);
+    }
+  }
+
+  int64_t left_padding() {
+    std::unique_lock<std::mutex> lock(mutex_);
+    return left_padding_;
+  }
+
+  void left_padding(int64_t size) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    left_padding_ = size;
+  }
+
+  int64_t right_padding() {
+    std::unique_lock<std::mutex> lock(mutex_);
+    return right_padding_;
+  }
+
+  void right_padding(int64_t size) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    right_padding_ = size;
+  }
+
+ protected:
+  // The background thread's main function
+  void WorkerLoop() {
+    std::unique_lock<std::mutex> lock(mutex_);
+    Status st;
+
+    while (true) {
+      if (please_close_) {
+        goto eof;
```

Comment:
Haven't seen `goto` in this codebase yet, but if it's the best way to write this code it is ok with me

Reply:
It's useful for cleanup code at the end of a function like this. I know C++ uses RAII when possible but it would be unwieldy here to define a local class just for this.

Reply:
No problem